### PR TITLE
EN-7697: Current network epoch provider component

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
Created a component that will hold the current epoch in the network. 
In the beginning, it relies on the epoch fetched from the epoch bootstrapper component. Later on, an epoch start meta block interceptor has been created and it will be used for requesting the network's epoch and will set the currentNetworkEpochProvider's epoch. To be decided when to use that interceptor

CC: @mihaib79

### CODE REVIEW

File: dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	cnrp.mutCurrentEpoch.RLock()
+	defer cnrp.mutCurrentEpoch.RUnlock()
+
+	lower := core.MaxInt(int(cnrp.currentEpoch)-cnrp.numActivePersisters+1, 0)
+	upper := cnrp.currentEpoch
+
+	return epoch >= uint32(lower) && epoch <= upper
+}
+
+// CurrentEpoch returns the
```
Comment (SebastianMarian) [inline dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go:43]: Finish the comment

File: dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+// NewCurrentNetworkEpochProvider will return a new instance of currentNetworkEpochProvider
+func NewCurrentNetworkEpochProvider(numActivePersisters int) *currentNetworkEpochProvider {
+	return &currentNetworkEpochProvider{
+		currentEpoch:        uint32(0),
+		numActivePersisters: numActivePersisters,
+	}
+}
+
+// SetCurrentEpoch will update the component's current epoch
+func (cnrp *currentNetworkEpochProvider) SetCurrentEpoch(epoch uint32) {
```
Comment (SebastianMarian) [inline dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go:24]: cnep ?

File: dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go (hunk 1 above)
Comment (bogdan-rosianu) [reply dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go:43]: done

File: dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go (hunk 2 above)
Comment (bogdan-rosianu) [reply dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go:24]: renamed

File: process/interceptors/epochStartMetaBlockInterceptor_test.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+		},
+	}
+	args.NumConnectedPeersProvider = &mock.MessengerStub{
+		ConnectedPeersCalled: func() []core.PeerID {
+			return make([]core.PeerID, 6) // 6 connected peers
+		},
+	}
+	args.ConsensusPercentage = 50 // 50% , so at least 3/6 have to send the same meta block
+
+	esmbp, _ := interceptors.NewEpochStartMetaBlockInterceptor(args)
```
Comment (SebastianMarian) [inline process/interceptors/epochStartMetaBlockInterceptor_test.go:133]: esmbi instead esmbp just to keep the same pattern in the whole file

Comment (bogdan-rosianu) [reply process/interceptors/epochStartMetaBlockInterceptor_test.go]: done

**Review by SebastianMarian — DISMISSED** [review]

File: cmd/node/main.go [hunk 4]
```diff
@@ -963,8 +964,23 @@ func startNode(ctx *cli.Context, log logger.Logger, version string) error {
 		return err
 	}
 
+	epochStartMetaBlockInterceptor, err := interceptors.NewEpochStartMetaBlockInterceptor(
+		interceptors.ArgsEpochStartMetaBlockInterceptor{
+			Marshalizer:               coreComponents.InternalMarshalizer,
+			Hasher:                    coreComponents.Hasher,
+			NumConnectedPeersProvider: networkComponents.NetMessenger,
+			ConsensusPercentage:       50,
```
Comment (SebastianMarian) [inline cmd/node/main.go:972]: You can define in core/constants.go this constant and use it here

File: dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go [hunk 5]
```diff
@@ -48,7 +82,99 @@ func (cnep *currentNetworkEpochProvider) CurrentEpoch() uint32 {
...
+
+	cnep.requestHandler.RequestStartOfEpochMetaBlock(uint32(math.MaxUint32))
+
+	return nil
+}
+
+func (cnep *currentNetworkEpochProvider) handlerEpochStartMetaBlock(_ string, _ []byte, data interface{}) {
+	metaBlock, ok := data.(*block.MetaBlock)
+	if !ok {
+		return
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go:181]: done. also added checks that the received meta block is of type `epoch start`

File: dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go [hunk 6]
```diff
@@ -37,7 +69,9 @@ func (cnep *currentNetworkEpochProvider) EpochIsActiveInNetwork(epoch uint32) bo
 	lower := core.MaxInt(int(cnep.currentEpoch)-cnep.numActivePersisters+1, 0)
 	upper := cnep.currentEpoch
 
-	return epoch >= uint32(lower) && epoch <= upper
+	isSynced := epoch >= uint32(lower) && epoch <= upper
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply dataRetriever/resolvers/epochproviders/currentNetworkEpochProvider.go]: outdated code. it is returned directly now

File: cmd/node/main.go (hunk 4 above)
Comment (bogdan-rosianu) [reply cmd/node/main.go:972]: done

**Review by SebastianMarian — APPROVED** [review]

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: process/interceptors/epochStartMetaBlockInterceptor.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+		return err
+	}
+
+	mbHash, err := core.CalculateHash(e.marshalizer, e.hasher, epochStartMb)
+	if err != nil {
+		return err
+	}
+
+	if !epochStartMb.IsStartOfEpochBlock() {
+		log.Debug("epochStartMetaBlockInterceptor-ProcessReceivedMessage: received meta block is not of "+
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply process/interceptors/epochStartMetaBlockInterceptor.go:72]: done

**Review by SebastianMarian — DISMISSED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**